packages/zoe/src/contracts/newSwap/getCurrentPrice.js

```diff
@@ -0,0 +1,260 @@
+ * @param {Brand} centralBrand
+ * @param {bigint} poolFeeBP - fee (BP) to the pool in every swap.
+ * @param {bigint} protocolFeeBP - fee (BP) charged in RUN to every swap.
+ */
+
+export const makeGetCurrentPrice = (
+  isSecondary,
+  isCentral,
+  getPool,
+  centralBrand,
+  poolFeeBP,
+  protocolFeeBP,
+) => {
+  const protocolFeeRatio = makeRatio(protocolFeeBP, centralBrand, BASIS_POINTS);
+  const halfPoolFeeBP = ceilDivide(poolFeeBP, 2);
+
+  const getPriceGivenAvailableInput = (amountIn, brandOut) => {
+    const { brand: brandIn } = amountIn;
+
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      // we'll charge the protocol fee before sending the remainder to the pool,
+      // so we need to subtract it from amountIn before getting a quote, and add
+      // it back before sending out the quote.
+      const prelimProtocolFee = multiplyBy(amountIn, protocolFeeRatio);
+      const poolAmountIn = amountMath.subtract(amountIn, prelimProtocolFee);
+      const price = getPool(brandOut).getPriceGivenAvailableInput(
+        poolAmountIn,
+        brandOut,
+        poolFeeBP,
+      );
+      const protocolFee = multiplyBy(price.amountIn, protocolFeeRatio);
+      const amountInFinal = amountMath.add(price.amountIn, protocolFee);
+
+      return {
+        amountIn: amountInFinal,
+        amountOut: price.amountOut,
+        protocolFee,
+      };
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      // We'll charge the protocol fee after getting the quote
+      const price = getPool(brandIn).getPriceGivenAvailableInput(
+        amountIn,
+        brandOut,
+        poolFeeBP,
+      );
+      const protocolFee = multiplyBy(price.amountOut, protocolFeeRatio);
+      const amountOutFinal = amountMath.subtract(price.amountOut, protocolFee);
+
+      return {
+        amountIn: price.amountIn,
+        amountOut: amountOutFinal,
+        protocolFee,
+      };
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
+      // We must do two consecutive getPriceGivenAvailableInput() calls,
+      // followed by a call to getPriceGivenRequiredOutput().
+      // 1) from amountIn to the central token, which tells us how much central
+      // would be provided for amountIn (centralAmount)
+      // 2) from centralAmount to brandOut, which tells us how much of brandOut
+      // will be provided (amountOut) as well as the minimum price in central
+      // tokens (reducedCentralAmount), then finally
+      // 3) call getPriceGivenRequiredOutput() to see if the same proceeds can
+      // be purchased for less (reducedAmountIn).
+
+      const brandInPool = getPool(brandIn);
+      const brandOutPool = getPool(brandOut);
+      const {
+        amountOut: centralAmount,
+      } = brandInPool.getPriceGivenAvailableInput(
+        amountIn,
+        centralBrand,
+        halfPoolFeeBP,
+      );
+
+      const firstDraftFee = multiplyBy(centralAmount, protocolFeeRatio);
```

Comment:
I'm probably just missing something, but it looks like this half of the fee is dropped. If it's not dropped, can you add a comment explaining where it is collected?